### DISCUSSION TITLE
How to analyze and reduce project building time?

### DISCUSSION THREAD

**Discussion by timsofteng** [vitejs/vite-plugin-react#84] [Q&A]
Hello. I'm trying to migrate from CRA to vite. It is typescript project.
When I start vite in development mode with this plugin the time printed in console is pretty low. It's around 1 second. But it takes about 20-30 seconds from starting dev server to moment when I see project on my screen.

How can I analize and reduce build time? 

P.s prod build takes even much more.

**Comment by ArnaudBarre** [comment]
You can run `vite --profile` and then press `p` when the page is loaded. This will dump a profile that can be loaded in chrome devtools. If the paths of your files in the project are safe to share, you can upload this profile so I can take a look.

You can also do the same thing with the current beta version of Vite 4.1, we made some performance improvements for import resolutions. It will be cool to see if the difference is significative or not.

If your project don't depend on some Babel specific option, you can try https://github.com/vitejs/vite-plugin-react-swc

You can also get some insights with https://github.com/antfu/vite-plugin-inspect

**Reply by timsofteng** [reply]
> You can run `vite --profile` and then press `p` when the page is loaded. This will dump a profile that can be loaded in chrome devtools. If the paths of your files in the project are safe to share, you can upload this profile so I can take a look.
> 

Unfortunately I cannot share this result because it is private. May you please suggest how to figure out with that dump or share some link about it?

> You can also do the same thing with the current beta version of Vite 4.1, we made some performance improvements for import resolutions. It will be cool to see if the difference is significative or not.
> 
Thanks! I'll try.

> If your project don't depend on some Babel specific option, you can try https://github.com/vitejs/vite-plugin-react-swc
> 
> You can also get some insights with https://github.com/antfu/vite-plugin-inspect
>
Project doesn't depend on babel but it is so tricky imports rules which you've described in readme to have hrm. Is it strict?

**Reply by ArnaudBarre** [reply]
You can read some information aout the devtools profiler here: https://developer.chrome.com/docs/devtools/performance

It's not strict, it just log a warning when the page reload because of it. This is almost identical between plugins and in v4 I will use the same logic as in the SWC plugin: https://github.com/vitejs/vite-plugin-react/pull/79

**Reply by timsofteng** [reply]
> You can read some information aout the devtools profiler here: https://developer.chrome.com/docs/devtools/performance
> 
Thanks.

> It's not strict, it just log a warning when the page reload because of it. This is almost identical between plugins and in v4 I will use the same logic as in the SWC plugin: #79
>
I mean this rules. https://github.com/ArnaudBarre/eslint-plugin-react-refresh
Looks like it is requirement for react-swc plugin only. We have a lot of such imports in our project unfortunately.

**Reply by ArnaudBarre** [reply]
This eslint plugin is not required, it just help you find places where HMR will fail & log a warning. Most people you the SWC plugin without this linter rule I think.

**Reply by timsofteng** [reply]
> This eslint plugin is not required, it just help you find places where HMR will fail & log a warning. Most people you the SWC plugin without this linter rule I think.

I mean import requirements. Hrm will not work with such imports if I've got correctly.

**Reply by ArnaudBarre** [reply]
The fast refresh requirements are almost the same on both plugins and other toolchain. When you use mixed exports, HMR will not work for this file and trigger HMR for more files or do a hard reload with a warning explaining why. But that's not required or more difficult than with others tools, on the opposite, it guides you to better HMR.
